### PR TITLE
Add a utility to assist in parsing text files with a line-oriented format.

### PR DESCRIPTION
**Pull request by ywkaras**

### CODE REVIEW

**Comment by zwoop** [comment]
Are there any details as to when / why someone would use this? With us moving to YAML, I imagine text file parsing is a thing of the past?

**Comment by ywkaras** [comment]
I guess we could always use YAML for all configs even all plugin configs.  It might be overkill for simpler plugin configs.

This code actually started out as something I used for a Yaolveroath-internal plugin.  I was mocking a proprietary library for an Au test, and the mocking code read in a script to control what was returned by each call to a mocked function.  I think it would have taken longer to code if I had used YAML.

**Comment by SolidWallOfCode** [comment]
If you're already using `TextView`, why not something like this
https://github.com/apache/trafficserver/blob/master/plugins/experimental/tls_bridge/tls_bridge.cc#L144

**Comment by ywkaras** [comment]
One difference is that it doesn't look like side comments are allowed in yours.

So you don't like this better?
```
using ts::LnParseIstream;

int line_no = 0, result = LnParseIstream::skipEmpty(ifs, line_no);
while (LnParseIsteam::EXTRA_FIELDS == result) {
  ++line_no;

  // Pick apart the line into the regular expression and destination service.
  std::string rxp, service;
  result = LnParseIstream(ifs, rxp, service);
  if (result != 2) {
    break;
  }
  this->load_pair(rxp, service, fp, line_no);
  result = LnParseIstream::skipEmpty(ifs, line_no);
}
if (result != LnParseIstream::END_OF_FILE) {
  TSError("[%s] %s, line %d in '%s'", PLUGIN_NAME, LnParseIstream::errStr(result), line_no, fp.c_str());
}
```

If any of the fields were integral or floating point, TextView would just run to its room crying.

I does kinda rub my fur the wrong way to read an entire file into memory to sequentially access it.  But that may just be cause my first programming class was on a machine with an 18-bit address space.

**Comment by SolidWallOfCode** [comment]
In practice on modern systems, configuration files are so small that it's much faster to read them in bulk and process in memory. As for comments, one can easily add a
```
if (line.ltrim_if(&isspace).empty || *line == '#') continue;
```
to allow comments that start with '#'.

For integers, I use `svtoi` and I haven't had to deal with floating point values in a configuration for quite a long time. But in practice, using `atof(token.data())` would work.

Also note that, in contrast to this last example code, there's only 1 memory allocation for the entire file, not two for every field (e.g. the `std::string` instances rxp, service) on every line.

**Comment by ywkaras** [comment]
OK now suppose that the line contained 10 parameters, 5 optional, instead of 2.  Which approach would be easier then?

**Comment by ywkaras** [comment]
Your comment about only one memory allocation reminded me of this line from the Woody Allen movie "Bananas":

 Fielding Mellish : I had a good relationship with my parents. They very rarely h-... I think they hit me once, actually, in my whole childhood. They, they, uh, started beating me on the 23rd of December in 1942, and stopped beating me in the late Spring of '44.

**Comment by SolidWallOfCode** [comment]
With regard to lines with varying numbers of arguments of different types, I think `TextView` is easier. I used it to do exactly that in the first pass on restructuring "remap.config" parsing. In fact, I would think `TextView` would be much easier if the type of the arguments varied as well, because if you pass references to them, that's a fixed ordering of types. I find it easier in such circumstances to have a more flexible API for working with the text.

I should note this is no accident - `TextView` derives directly from requirements I had while writing text parsers for TS, borrowing concepts from other languages that had proven useful (e.g., `trim`).

**Comment by ywkaras** [comment]
But this isn't a choice between having just one or the other, it's a choice of having just one or both.  It's a reasonable position to say I should not have bothered to write this.  But I did, so that's where we are now.  I think your key point that I'm not eliminating much per field, a prefix() call, a trim call, maybe a conversion function call.  But if this utility is available, why do those extra steps if you don't have to?  Consider the extra steps with TextView if you want to allow for quoted string fields, or optionally quoted string fields.  Also, I think I could add line continuation to LnParseIstream fairly easily.

**Comment by SolidWallOfCode** [comment]
There's a balance between a reasonably sized API and having a method for everything you might ever want to do while parsing. One thing I like about where `TextView` is on that continuum is it's easy to tweak to handle slightly different cases, e.g. different separators, nested separators (e.g. "key=value" pairs in comma separated lists). Before the switch to YAML, this was a critical feature in dealing with common ATS configuration files. A related aspect is this enables me to provide much more detailed error messages when things are not as expected.

There's overhead to every additional library, both in maintenance and understanding. My natural bias is `TextView` is in the sweet spot for balancing that and utility, primarily because I built it to be where I perceived that spot to be. Your mileage may vary. I personally wouldn't use this because I find the additional utility to be worth less than the flexibility of using only `TextView` in the situations that come up in working on TS. That might be different in other projects, or other opinions, but that's my view.

**Comment by ywkaras** [comment]
A prefix custom field could be added to help deal with key=value, key:value and such if those are common cases.

File: src/tscpp/util/unit_tests/test_LnParseIstream.cc [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    REQUIRE(1 == line_num);
+  }
+  {
+    std::stringstream is(" ");
+
+    int catch_sucks = LnParseIstream::STREAM_ERROR;
+    REQUIRE(LnParseIstream(is) == catch_sucks);
+  }
+  {
+    std::stringstream is("5 \"unterminated string\n");
```
Comment (ywkaras) [inline src/tscpp/util/unit_tests/test_LnParseIstream.cc:174]: Oops parsing is missing.

**Comment by ywkaras** [comment]
Putin has declined my request to have FSB agents surreptitiously sprinkle plutonium on @SolidWallOfCode 's food, this this PR can never be merged.

**Pull request closed without merging**